chore(ci): Add dev tests with master dependencies

kramaranya opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines: https://www.kubeflow.org/docs/about/contributing
2. To know more about Kubeflow SDK, check the developer guide:
    https://github.com/kubeflow/sdk/blob/main/CONTRIBUTING.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:
I added a test-python-dev target and changed the unit test CI jobs with a dev/prod matrix

Dev leg installs the dev dependency group, which now pulls `kubeflow_trainer_api` from the master branch. Prod leg will run with the packages published to PyPI (once API models are available there)

I also updated the CI job to run unit tests for two Python vesions -- `3.9` and `3.11`

I removed flake8 since ruff replaced it

/assign @andreyvelich @astefanutti @Electronic-Waste @tenzen-y @briangallagher

**Checklist:**

- [x] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

Electronic-Waste reviewed (commented):
@kramaranya Thanks for this!
/lgtm

kramaranya commented:
@astefanutti @Electronic-Waste thanks for your review!
I’ve updated the CI to run unit tests on Python 3.9 and 3.11, and removed flake8 because ruff replaced it in #38 
Could you please take a look again?

andreyvelich commented on python/pyproject.toml:
```diff
@@ -55,6 +55,7 @@ ignore = ["E203"]
 [dependency-groups]
 dev = [
     "ruff>=0.12.2",
+    "kubeflow_trainer_api@git+https://github.com/kubeflow/trainer.git@master#subdirectory=api/python_api",
```
Is there any use-case when users would like to install Kubeflow SDK with the latest module package ?
IIUC, the dependency-groups is never published to PyPI by hatch, so users can't install.
Thoughts @astefanutti @szaher @briangallagher @kramaranya @eoinfennessy ?

kramaranya replied:
Yeah, I've been thinking about adding this to `project.optional-dependencies` as well, but since we don't update API models that often, it makes sense to me to keep models from master available only for testing. However, I'm open to other suggestions if you see it differently

andreyvelich replied:
From my point of view, since we allow to install control plane with the latest changes, users might want to experiment with it during the development: https://www.kubeflow.org/docs/components/trainer/operator-guides/installation/#installing-the-kubeflow-trainer-controller-manager

This gives opportunities for them to quickly adopt newer API changes into their systems.

kramaranya replied:
Makes sense, let me update it

andreyvelich commented on Makefile:
```diff
@@ -67,9 +67,20 @@ uv-venv:
 	fi
 
  # make test-unit will produce html coverage by default. Run with `make test-unit report=xml` to produce xml report.
+.PHONY: test-python-dev
+test-python-dev: uv-venv
+	@uv pip install --project $(PY_DIR) --group dev "$(PY_DIR)[test]"
+	@uv run coverage run --source=kubeflow.trainer.api.trainer_client,kubeflow.trainer.utils.utils -m pytest ./python/kubeflow/trainer/api/trainer_client_test.py
+	@uv run coverage report -m kubeflow/trainer/api/trainer_client.py kubeflow/trainer/utils/utils.py
+ifeq ($(report),xml)
+	@uv run coverage xml
+else
+	@uv run coverage html
+endif
```
Can we run tests over the master modules for now ?
Since we can't introduce any breaking changes to the Kubernetes API, I don't see a need to run tests against released modules.

kramaranya replied:
> Since we can't introduce any breaking changes to the Kubernetes API

Hmm, why can't we?

andreyvelich replied:
According to Kubernetes API best practices, if we introduce any breaking changes to the existing API: `v1alpha1`, we have to create the newer API version, for instance: `v1alpha2`: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

cc @tenzen-y @astefanutti 

In the future, we should discuss how to design our SDK to make it compatible with supported API versions, but we are not there yet.

andreyvelich commented on .github/workflows/test-python.yaml:
```diff
@@ -5,25 +5,52 @@ on:
   push:
     branches: [main]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
IIRC, we discussed before that this part is not needed given that we don't have limitations on concurrent runs due to GitHub enterprise.

astefanutti replied:
A use case is re-pushing the branch for the PR while the workflow is running. Even with no limitations, is that useful to keep the existing run going?

andreyvelich replied:
Might be useful to compare results, but I am fine to cancel them for now.
@kramaranya Please can you apply those changes to all GitHub action tests in that case ?

andreyvelich commented on .github/workflows/test-python.yaml:
```diff
@@ -5,25 +5,52 @@ on:
+          flag-name: ${{ matrix.dependency-group.name }}-py${{ matrix.python-version }}
+          parallel: true
+
+  coverage:
```
Don't we run coverage as part of the previous job in Upload coverage to Coveralls step ?

andreyvelich replied:
As I can see, it shows the report on the PRs: https://github.com/kubeflow/sdk/pull/56#issuecomment-3144804385

andreyvelich reviewed (commented):
Thank you for this @kramaranya!

andreyvelich commented on python/pyproject.toml:
Let's also migrate to the official package as part of this PR: https://pypi.org/project/kubeflow-trainer-api/

kramaranya commented:
@andreyvelich @astefanutti thank you for the review!
I should have addressed all your comments and imported `kubeflow-trainer-api` from PyPI https://github.com/kubeflow/sdk/blob/e84a54fcd70bc4aeda3008e76fdee9d104d888f7/python/pyproject.toml#L31.
Please let me know if there's anything else I should update!

astefanutti commented:
/lgtm

Thanks @kramaranya!

andreyvelich commented on CONTRIBUTING.md:
````diff
@@ -28,6 +28,12 @@ cd python
+
 Install development tools:
 ```sh
 pip install pytest black isort flake8 coverage pre-commit
````
I guess, we can remove it since we have everything in dev, right ?

kramaranya replied:
yep, updated in https://github.com/kubeflow/sdk/pull/55/commits/cb1c4dfd862f92cfaa6bafee308f30dcc18821a3

andreyvelich commented on Makefile:
```diff
@@ -66,10 +66,10 @@ uv-venv:
 .PHONY: test-python
 test-python: uv-venv
-	@uv pip install "./python[test]"
+	@uv pip install --project $(PY_DIR) --group dev "$(PY_DIR)[test,dev]"
```
Please can we [move `ruff`](https://github.com/kubeflow/sdk/blob/e84a54fcd70bc4aeda3008e76fdee9d104d888f7/python/pyproject.toml#L60) and [`test` ](https://github.com/kubeflow/sdk/blob/e84a54fcd70bc4aeda3008e76fdee9d104d888f7/python/pyproject.toml#L34) packages to `dev` too ?

In that case we have all tools required for development in one place

kramaranya replied:
Sure, shall we also add `pre-commit` package to `dev`?

kramaranya commented:
@andreyvelich thank you for the review! Let me please know if there's anything else I should update!

kramaranya commented on Makefile:
```diff
@@ -66,10 +66,10 @@ uv-venv:
 .PHONY: test-python
 test-python: uv-venv
-	@uv pip install "./python[test]"
+	@uv pip install --project $(PY_DIR) -e "$(PY_DIR)[dev]"
```
done in https://github.com/kubeflow/sdk/pull/55/commits/542ddcd946eab05ff4c4cf831d6f30cccb82ff3c

andreyvelich reviewed (commented):
Thanks for this @kramaranya!
/lgtm
/approve
